### PR TITLE
Support for Vault AppRole Auth method

### PR DESCRIPTION
**Pull request by kamalakarp**
Support for Vault AppRole Auth method

### CODE REVIEW

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/HttpRequestConfigTokenProvider.java [hunk 1]
```diff
@@ -42,12 +42,11 @@ public String getToken() {
 		}
 
 		String token = request.getHeader(ConfigClientProperties.TOKEN_HEADER);
-		if (!StringUtils.hasLength(token)) {
```
[the remark this answers is not in the document]
Comment (spencergibb) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/HttpRequestConfigTokenProvider.java]: But this breaks current functionality. When would this provider be used but need to be ignored?

[the remark this answers is not in the document]
Comment (kamalakarp) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/HttpRequestConfigTokenProvider.java]:
```
spring:
  profiles:
    active: vault, git
```
Whenever there an active profile with vault, and if the token is empty - that's when AppRole auth gets used.

File: spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServicePropertySourceLocator.java [hunk 2]
```diff
@@ -247,6 +250,11 @@ private Environment getRemoteEnvironment(RestTemplate restTemplate,
 				if (StringUtils.hasText(token)) {
 					headers.add(TOKEN_HEADER, token);
 				}
+				// Sending the client headers to server "headers:"
+				ArrayList<String> keyList = new ArrayList<String>(clientHeaders.keySet());
```
Comment (spencergibb) [inline spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServicePropertySourceLocator.java]: This is already happening below, see `GenericRequestHeaderInterceptor`and `getSecureRestTemplate()`

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/HttpRequestConfigTokenProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (spencergibb) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/HttpRequestConfigTokenProvider.java]: Then the condition needs to change. If the other headers are in the request, don't throw the exception.

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/VaultEnvironmentRepository.java [hunk 3]
```diff
@@ -204,23 +230,37 @@ private void addProfiles(List<String> contexts, String baseContext,
...
 
-	private String getToken() {
-		String token = tokenProvider.getToken();
-		if (!StringUtils.hasLength(token)) {
-			throw new IllegalArgumentException(
-					"A Vault token must be supplied by a token provider");
+	private String getToken(HttpHeaders vaultHeaders) {
+		String vaultToken = tokenProvider.getToken();
+		if (!StringUtils.hasLength(vaultToken)) {
+			HttpServletRequest servletRequest = this.request.getIfAvailable();
```
Comment (scottfrederick) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/VaultEnvironmentRepository.java]: All of the token retrieval logic was previously moved from `VaultEnvironmentRepository` to the `ConfigTokenProvider` implementations to better support the embedded config server use case (where there is no `HttpServletRequest` to get auth info from). This change re-introduces the `HttpServletRequest` requirement into `VaultEnvironmentRepository`, which is a step backwards IMO.

Comment (spencergibb) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/VaultEnvironmentRepository.java]: Good catch. There will need to be an analogous interface for this.

**Comment by spencergibb** [comment]
Superseded by #1526

**Pull request closed without merging**